#2534 UID Mapping

vaclavbasniar opened the pull request:
[//]: # (Request Template Witsml Explorer)

[//]: # (Thank you for contributing to WITSML Explorer! Before submitting this PR, please fill in the following template.)

## Fixes

[//]: # (Write the GitHub issue number starting with #, or, for Equinor only, the Jira issue number starting with WE-)

This pull request fixes issue #2534

## Description

[//]: # (Please include a written summary of the changes and which issue is fixed. List any dependencies that are required for this change._)

- added modal
- added storage
- updated copy and show on server functionality
- removed connected server from list of servers in context menus
- unit tests

## Type of change

[//]: # (Mark any of the types of change that apply.)

* [ ] Bugfix
* [x] New feature (non-breaking change which adds functionality)
* [x] Enhancement of existing functionality
* [x] Breaking change (fix or feature that would cause existing functionality to not work as expected)
* [x] This change requires a documentation update

## Impacted Areas in Application

[//]: # (List general components of the application that this PR will affect)

* [x] Frontend
* [x] API
* [ ] WITSML
* [ ] Desktop
* [ ] Other (please describe)

## Checklist:

[//]: # (Please tick all the boxes or remove the ones that aren't needed and explain why)

*Communication*
* [ ] I have made corresponding changes to the documentation
* [ ] PR affects application security

*Code quality*
* [x] I have self-reviewed my code
* [x] No new warnings are generated

*Test coverage*
* [x] New code is covered by passing tests

## Further comments

[//]: # (If this is a relatively large or complex change, kick off the discussion by explaining why you chose the solution you did and what alternatives you considered, etc...)

vaclavbasniar committed:
merged main

vaclavbasniar committed:
merged main

eliasbruvik commented on Src/WitsmlExplorer.Api/Models/UidMapping.cs:
```diff
@@ -0,0 +1,81 @@
+        public string TargetWellboreId { get; set; }
+
+        [JsonPropertyName("username")]
+        public string Username { get; set; }
```
What is the intention of _Username_ for a mapping? Can't see that it's used anywhere in the frontend. Is it to keep a reference to the person who last created/updated a mapping?

vaclavbasniar replied:
Yes, as a reference to a person who did the mapping. Will be shown in Uid Mapping Overview #2568

eliasbruvik commented on Src/WitsmlExplorer.Api/Services/UidMappingService.cs:
```diff
@@ -0,0 +1,189 @@
+        public async Task<ICollection<UidMapping>> QueryUidMapping(UidMappingDbQuery query)
+        {
+            if ((query.SourceWellId.IsNullOrEmpty() && query.SourceWellboreId.IsNullOrEmpty())
+                || (!query.SourceWellId.IsNullOrEmpty() && !query.SourceWellboreId.IsNullOrEmpty()))
```
Noticed that you don't check the server Guid's like you do in the Verify method in the handler. Is this intentional? Is there any reason why we can't validate it in the handler like we do for create/update?

vaclavbasniar replied:
added validation

eliasbruvik commented on Src/WitsmlExplorer.Api/Services/UidMappingService.cs:
```diff
@@ -0,0 +1,189 @@
+using System.Threading.Tasks;
+
+using Microsoft.AspNetCore.Http;
+using Microsoft.AspNetCore.Mvc;
```
There are several unused directives in most of the .cs files. Could you remove these?

vaclavbasniar replied:
removed

eliasbruvik commented on Src/WitsmlExplorer.Frontend/services/uidMappingService.tsx:
```diff
@@ -0,0 +1,73 @@
+    }
+  }
+
+  public static async queryRemoveUidMapping(
```
Shouldn't this be _removeUidMapping_ without _query_?

vaclavbasniar replied:
changed

eliasbruvik commented on Src/WitsmlExplorer.Frontend/services/uidMappingService.tsx:
```diff
@@ -0,0 +1,73 @@
+    }
+  }
+
+  public static async queryRemoveUidMapping(
```
Doesn't look like this is currently used. Do you have any plans on supporting removal of mappings? What would currently happen if you incorrectly map something?

vaclavbasniar replied:
Delete is going to be implemnted in Uid Mapping Overview #2568

eliasbruvik replied:
Makes sense, hadn't noticed that issue yet.

eliasbruvik commented on Src/WitsmlExplorer.Api/Services/UidMappingService.cs:
Is the mapping bi-directional? If so, how is this currently achieved? If not, is there any reason why it shouldn't be?

vaclavbasniar replied:
We had a discussion about this with Aker and it would be better to keep it uni-directional for now to have full control over the mapping.

eliasbruvik commented on Src/WitsmlExplorer.Api/Services/UidMappingService.cs:
```diff
@@ -0,0 +1,189 @@
+
+        public async Task<UidMapping> CreateUidMapping(UidMapping uidMapping, HttpContext httpContext)
+        {
+            var mappings = await _mappingRepository.GetDocumentAsync(new(uidMapping.SourceServerId, uidMapping.TargetServerId));
```
Tested it with LiteDB, but got this error. It needs to work with every database we support (LiteDB, CosmosDB & MongoDB). 
![image](https://github.com/user-attachments/assets/774a0e5e-b318-4e18-a348-da2225657f76)

vaclavbasniar replied:
Replaced complex object with serialized JSON

eliasbruvik reviewed (changes requested):
Sorry for taking a while. Started on a review, but not finished yet. Anyways, added some comments on things that can be improved, and some questions to help me understand the logic.

eliasbruvik commented on Src/WitsmlExplorer.Frontend/components/ContextMenus/ContextMenuUtils.tsx:
```diff
@@ -87,56 +90,96 @@ export const onClickShowObjectOnServer = async (
+      targetServerId: targetServer.id
+    };
+
+    const mappings = await UidMappingService.queryUidMapping(dbQuery);
```
Is it intentional to only use the mapping for some of the objects? I think you probably want to apply this to all the if/else cases.

vaclavbasniar replied:
It was intentional at first, however it could be implemented for all cases. Since this is just show on server and in the case there is no corresponding object on wellbore with matching UID on target server, information that the object doesn't exist is shown.

eliasbruvik replied:
I mean `if (objectType === ObjectType.Log) {...` and `} else if (isExpandableGroupObject(objectType)) {...`. Currently you are not using the mapping for showOnServer for these objects: 
![image](https://github.com/user-attachments/assets/2d0246bb-9ee9-401e-a58b-db548f3605d7)

vaclavbasniar replied:
added

eliasbruvik commented on Src/WitsmlExplorer.Frontend/components/ContextMenus/ContextMenuUtils.tsx:
```diff
@@ -87,56 +90,96 @@ export const onClickShowObjectOnServer = async (
+  } else {
+    targetWellId = wellbore.wellUid;
+    targetWellboreId = wellbore.uid;
+  }
```
There are several occasions where logic similar to the one you added above is used. Can you create an async function that accepts the necessary parameters and returns the mapping id's if existant, or the input id's if not?

vaclavbasniar replied:
added function for most of the cases

vaclavbasniar committed:
#2534 PR changes
- refactoring
- removed unused usings
- added missing objects mappings for show on server
- added missing validation
- changed complex key to serialized JSON (LiteDB doesn't support complex objects as key)
- updated tests

vaclavbasniar committed:
#2534 PR changes
- refactoring
- removed unused usings
- added missing objects mappings for show on server
- added missing validation
- changed complex key to serialized JSON (LiteDB doesn't support complex objects as key)
- updated tests

eliasbruvik commented on Src/WitsmlExplorer.Frontend/components/ContextMenus/CopyToServer.tsx:
```diff
@@ -31,38 +33,63 @@ export const onClickCopyToServer = async (
+    targetServerId: targetServer.id
   };
 
+  const mappings = await UidMappingService.queryUidMapping(dbQuery);
```
Can't getTargetWellboreID also be used here? Not tested, but something like

```
const sourceWellboreName = toCopy[0].wellboreName;

  const { targetWellId, targetWellboreId } = await getTargetWellboreID({
    sourceServerId: sourceServer.id,
    sourceWellId: sourceWellUid,
    sourceWellboreId: sourceWellboreUid,
    targetServerId: targetServer.id
  });

  let wellbore: Wellbore;
  try {
    wellbore = await WellboreService.getWellbore(
      targetWellId,
      targetWellboreId,
      null,
      targetServer
    );
  } catch {
    return; // Cancel the operation if unable to authorize to the target server.
  }

  const targetWellboreRef: WellboreReference = {
    wellUid: targetWellId,
    wellboreUid: targetWellboreId,
    wellName: wellbore?.wellName ?? sourceWellName,
    wellboreName: wellbore?.name ?? sourceWellboreName
  };
```

vaclavbasniar replied:
Yes, I'll check that again. However, I guess this would probably need a little bit more refactoring.

eliasbruvik commented on Src/WitsmlExplorer.Frontend/components/ContextMenus/CopyToServer.tsx:
If I map a wellbore to a wellbore on another server, and then delete the target wellbore, the mapping will exist but not the target wellbore. If I then try to use the copyToServer-functionality it asks me to create the target wellbore as expected. However, it fails with this message:
![image](https://github.com/user-attachments/assets/997be403-2617-4c0c-b9c2-bf02649cb5b9)

I would expect either of these (maybe this should be discussed?):
* Any mapping that exists for a wellbore that is deleted should also be removed. If this is chosen, we need to take into account what should happen if the wellbore is deleted from the query view or another application (when we don't know that it has been deleted).
* CopyToServer will create the wellbore with the mapped uid, and then copy the object over (without any errors).

vaclavbasniar replied:
Yes, that's a good point, that hasn't been discussed. I think, the first solution is correct. We should probably discuss this in weekly meeting.

vaclavbasniar committed:
Merge branch 'main' into features/2534_wellbore_uid_mapping

vaclavbasniar committed:
Merge branch 'features/2534_wellbore_uid_mapping' of https://github.com/vaclavbasniar/witsml-explorer into features/2534_wellbore_uid_mapping

vaclavbasniar committed:
Merge branch 'main' into features/2534_wellbore_uid_mapping

# Conflicts:
#	Src/WitsmlExplorer.Frontend/components/ContextMenus/FluidContextMenu.tsx
#	Src/WitsmlExplorer.Frontend/components/ContextMenus/LogCurveInfoContextMenu.tsx
#	Src/WitsmlExplorer.Frontend/components/ContextMenus/LogsContextMenu.tsx
#	Src/WitsmlExplorer.Frontend/components/ContextMenus/ObjectMenuItems.tsx
#	Src/WitsmlExplorer.Frontend/components/ContextMenus/ObjectsSidebarContextMenu.tsx
#	Src/WitsmlExplorer.Frontend/components/ContextMenus/TrajectoryStationContextMenu.tsx
#	Src/WitsmlExplorer.Frontend/components/ContextMenus/TubularComponentContextMenu.tsx
#	Src/WitsmlExplorer.Frontend/components/ContextMenus/WbGeometrySectionContextMenu.tsx
#	Src/WitsmlExplorer.Frontend/components/ContextMenus/WellContextMenu.tsx
#	Src/WitsmlExplorer.Frontend/components/ContextMenus/WellboreContextMenu.tsx

vaclavbasniar committed:
fix after merge

vaclavbasniar committed:
#2534 UID Mapping
* bug fixes
* automatic removal of UID mapping on well/wellbore delete
* automatic cleanup of UID mappings pointing to deleted well/wellbores on copy to
* unit tests update

vaclavbasniar committed:
#2534 added missing readonly modifiers

vaclavbasniar committed:
#2534 removed unused usings

eliasbruvik commented on Src/WitsmlExplorer.Frontend/components/Modals/WellboreUidMappingModal.tsx:
Tested how this would work on a production server that has a lot of wells/wellbores. It completely freezes the application for about 30 sec whenever I make a small change to the search. 

It's a bit laggy (~1 sec) when changing a search with few matches, and very laggy when there are a lot of results, so this might mostly be a rendering issue. Can you try to add a small debounce to the input to avoid unnecessary full re-renders, and to virtualize the treeview?

vaclavbasniar replied:
Yes, I believe, that it's the rendering issue. I'll add the virtualization and debounce.

vaclavbasniar replied:
Is there a possibility we could get a test server with such a setup - lot of well, wellbores, data in general? I think it could be beneficial also for future development cases.

eliasbruvik replied:
I agree that it would be beneficial, but it hasn't been easy to get access to "public" test servers in the past. We had some discussions in the past on getting more realistic datasets over to the public servers, but that has been downprioritized for a while. If we want more data, I think we would have to generate a lot of wells/wellbores on one of the public servers ourselves.

vaclavbasniar replied:
Added debounce, virtualization and some small performance improvements.

vaclavbasniar committed:
Merge branch 'main' into features/2534_wellbore_uid_mapping

vaclavbasniar committed:
#2534
- optimization
- virtualization
- refactoring
- debounce

vaclavbasniar committed:
#2534
- removed virtualizer
- added search result limit

eliasbruvik reviewed (dismissed):
*Reviewer checklist*
- [x] Read issue and PR
- [x] Pulled branch and manually tested
- [x] Verify that PR resolves issue
- [x] Reviewed the code

vaclavbasniar committed:
#2534
- fixed well node toggling
